File: main/delete-node-in-a-bst/delete-node-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode deleteNode(final TreeNode root, final int key) {
        if (root == null) return null;

        if (key != root.val) {
            if (key < root.val)
                root.left = deleteNode(root.left, key);
            else
                root.right = deleteNode(root.right, key);

            return root;
        }

        if (root.left == null) return root.right;
        if (root.right == null) return root.left;

        final TreeNode node = extractNextBelow(root);
        reparent(root, node);
        return node;
    }

    private static TreeNode extractNextBelow(final TreeNode root) {
        TreeNode node = root.right;

        if (node.left == null) {
            root.right = node.right;
        } else {
            while (node.left.left != null) node = node.left;
            final TreeNode parent = node;
            node = node.left;
            parent.left = node.right;
        }

        node.right = null; // makes debugging easier
        return node;
    }

    private static void reparent(final TreeNode oldParent,
                                 final TreeNode newParent) {
        newParent.left = oldParent.left;
        newParent.right = oldParent.right;

        oldParent.left = oldParent.right = null; // makes debugging easier
    }
}

File: main/delete-node-in-a-bst/delete-node-in-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} key
 * @return {TreeNode}
 */
function deleteNode(root, key) {
    'use strict';

    if (!root) return null;

    if (key < root.val) {
        root.left = deleteNode(root.left, key);
        return root;
    }

    if (root.val < key) {
        root.right = deleteNode(root.right, key);
        return root;
    }

    if (root.val !== key) {
        throw new Error('incommensurable keys');
    }

    // If it has at most one child, just replace it with its child (or null).
    if (!root.right) return root.left;
    if (!root.left) return root.right;

    // Its right child has no left child, so attach its left child as the left
    // child of its right child.
    if (!root.right.left) {
        root.right.left = root.left;
        return root.right;
    }

    // Find the minimal node in its right subtree.
    let parent = root.right;
    let child = parent.left;
    while (child.left) {
        parent = child;
        child = child.left;
    }

    // Replace it with that minimal node. Very carefully.
    parent.left = child.right;
    child.right = root.right;
    child.left = root.left;
    return child;
}
